### DISCUSSION TITLE
Airflow EmrServerlessStartJobOperator return success even serverless job fails

### DISCUSSION THREAD

**Discussion by dviru** [apache/airflow#28294] [Q&A] [answered]
Hi Team, I am using EmrServerless Operator, Sometime my Job will fail but in airflow I am seeing task in success state, when I debugging suspecting some code issue in EmrServerlessHook. Please find my observation as given below.

<img width="1067" alt="image" src="https://user-images.githubusercontent.com/10843400/206966496-214c8913-6fef-4704-9870-0d16eebaa302.png">

At line number: 145 it will check the current job state is not exist in desired_state **(This value will be EmrServerlessJobSensor.TERMINAL_STATES which means {'SUCCESS', 'FAILED', 'CANCELLING', 'CANCELLED'} refer code [click here](https://github.com/apache/airflow/blob/2.4.3/airflow/providers/amazon/aws/operators/emr.py#L653) )** , assume at line-number 152 we got failed state of the current job, in that case line-number 145 condition not satisfied and it won't enter the while loop and soft exit happens, So from this methods won't raise any exception at the time of Failed state and it is considering as success.

Please let me know if my understanding is wrong

**Comment by potiuk** [comment]
@vincbeck @ferruzzi @o-nikolas -> likely something for you to take a look

**Reply by ferruzzi** [reply]
I'll take a look; Also going to ping @syedahsn since he's the most familiar with EMR Serverless

**Reply by ferruzzi** [reply]
Nevermind, vincbeck beat me to it :P

**Comment by Taragolis** [comment]
@dviru which version of Amazon provider are you use? Because your links refers to Airflow 2.4.3 tag, however [providers release cycle slightly different](https://github.com/apache/airflow#release-process-for-providers) and depend on minimum Airflow version

I check latest version of [amazon provider 6.2.0](https://github.com/apache/airflow/blob/providers-amazon/6.2.0/airflow/providers/amazon/aws/operators/emr.py) and it's code quite different.

Could you confirm that this issue exists or not in latest version of provider? Because it also a chance that this issue already fixed, [see amazon provider changelog](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/index.html#changelog)

**Reply by vincbeck** [reply]
Good catch! I think the bug is still there. The code had moved since then but remains the same, it is now available here: https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/utils/waiter.py#L57. I created an issue for it, see #28691

**Comment by vincbeck** [comment]
I think there is a bug in the `waiter` function but also in the operator. The `waiter` function has `Failed` in `desired_state`, thus this is expected this function does not throw an error when the state is failed. I think the bug is [here](https://github.com/apache/airflow/blob/2.4.3/airflow/providers/amazon/aws/operators/emr.py#L653) when `desired_state` should only be [SUCCESS_STATE](https://github.com/apache/airflow/blob/c53a3e153f0ab5ca933a94adc01dc6314ea8d4d1/airflow/providers/amazon/aws/sensors/emr.py#L132). I'll create a PR for both issues

**Comment by vincbeck** [comment, marked as answer]
It's actually already fixed. [See here](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/operators/emr.py#L878). @dviru please upgrade to the latest Amazon provider package to fix the issue.

The `waiter` function has still a minor bug and needs refactoring to me, I'll create a PR for it (associated to the issue I created) but this is not related to this bug. Upgrading to the latest Amazon provider package fixes the bug
